whirlpool/eventsocket.py

```diff
@@ -62,95 +62,96 @@ async def _recv_msg(self, websocket: aiohttp.ClientWebSocketResponse):
         return msg
 
     async def _run(self):
-        if not self._running:
-            return
-        timeout = aiohttp.ClientTimeout(total=None, connect=60, sock_connect=60)
-
-        try:
-            LOGGER.debug(f"Connecting to {self._url}")
-            async with self._session.ws_connect(
-                self._url,
-                timeout=timeout,
-                autoclose=True,
-                autoping=True,
-                heartbeat=45,
-            ) as ws:
-                self._websocket = ws
-                await self._send_msg(ws, self._create_connect_msg())
-                await self._recv_msg(ws)
-                await self._send_msg(ws, self._create_subscribe_msg())
-                await self._con_up_listener()
-
-                self._reconnect_tries = RECONNECT_COUNT
-
-                while not ws.closed:
-                    msg = await self._recv_msg(ws)
-                    if not msg:
-                        continue
-                    if msg.type == aiohttp.WSMsgType.ERROR:
-                        LOGGER.error("Socket message error")
-                        break
-                    if msg.type in [
-                        aiohttp.WSMsgType.CLOSE,
-                        aiohttp.WSMsgType.CLOSING,
-                        aiohttp.WSMsgType.CLOSED,
-                    ]:
-                        LOGGER.warn(
-                            f"Stopping receiving. Message type: {str(msg.type)}"
-                        )
-
-                        if (
-                            not self._auth.is_access_token_valid()
-                            or msg.data == WS_STATUS_UNAUTHORIZED
-                        ):
-                            LOGGER.debug("auth key expired, doing reauth now")
-                            await self._auth.do_auth()
-
-                        elif msg.data == WS_STATUS_GOING_AWAY:
+        while self._running:
+            timeout = aiohttp.ClientTimeout(total=None, connect=60, sock_connect=60)
+
+            try:
+                LOGGER.debug(f"Connecting to {self._url}")
+                async with self._session.ws_connect(
+                    self._url,
+                    timeout=timeout,
+                    autoclose=True,
+                    autoping=True,
+                    heartbeat=45,
+                ) as ws:
+                    self._websocket = ws
+                    await self._con_up_listener()
```

Comment:
why move this up?

Reply:
This is the loop change - moved to a while loop, removed the thread restart logic - aka just keep looping until `not self._running`  There's really no need to let the thread die then restart it.
The `stop()` logic sets self._running = False to cause this thread to end.

Reply:
Sorry, I moved that to trigger a login if it gets a 401.  With it in the original location, we can get an error and never try a login.

Reply:
Sorry, I am not sure I understand. I think `_con_up_listener()` should only be called after the connect and subscribe messages are sent, since that is effectively when the connection is ready.
This will loop if we get an error, right?

Reply:
`_con_up_listener()' is `fetch_data()` - https://github.com/abmantis/whirlpool-sixth-sense/blob/26d8cecbec6f0ff2e6a83efa8db54b3c8197a3d3/whirlpool/appliance.py#L93-L108
Calling fetch data first, makes sure we are logged in before attempting to register the callacks via the code that follows.  I witnessed this occurring so I moved this up.  I'm not sure how we actually became logged out before attempting to register, but it happened.

Reply:
> `_con_up_listener()` is `fetch_data()` -

Yes, I know. But the PoV of this class (Eventsocket), `_con_up_listener` can be anything. On Eventsocket we should not assume that the callback does "X".
> 
> https://github.com/abmantis/whirlpool-sixth-sense/blob/26d8cecbec6f0ff2e6a83efa8db54b3c8197a3d3/whirlpool/appliance.py#L93-L108
> 
> Calling fetch data first, makes sure we are logged in before attempting to register the callbacks via the code that follows. I witnessed this occurring so I moved this up. I'm not sure how we actually became logged out before attempting to register, but it happened.
> 
> Calling `fetch_data()` after the change of state is registered was intended to make sure the state machine is updated before switch to 100% change of state. By calling it "just before" registering there is a small chance we miss a state change, but it's better than not registering due to a login error and being 100% off until the integration is reloaded.

But Eventsocket does not know that, so the callback should be generic anyway. Otherwise, if in the future `fetch_data()` changes and does not login for some reason, then this assumption will be broken. Also, someone working on this later will also not know that we had this assumption here, by looking at Eventsocket.
So, if we need to ensure that auth is done before, that should be handled entirely here without relying on the callback to do it, IMO.

Reply:
So what's the answer? Add a login function here if we get a 401?  It can't just "die" because were not logged at this point.

Reply:
I've added a `check_login` function & callback - is this what you had in mind?